### PR TITLE
synaptics-rmi: support latest firmware bootloader (~Ver 10.0) and usb…

### PR DESCRIPTION
**Pull request by vhuag**
This commit contains the changes related to latest BL version.

1. Supports the containerIDs and partitionIDs in latest BL version.
2. Supports signature verification.
3. Supports USBHID devices for development and debugging.

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

File: plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c [hunk 1]
```diff
@@ -93,6 +90,8 @@ typedef enum {
 	RMI_FIRMWARE_CONTAINER_ID_EXTERNAL_TOUCH_AFE_CONFIG,
 	RMI_FIRMWARE_CONTAINER_ID_UTILITY,
 	RMI_FIRMWARE_CONTAINER_ID_UTILITY_PARAMETER,
+	/* Reserved : 24 ~ 26 */
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c:93]: lowercase `Reserved` please

File: plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c [hunk 2]
```diff
@@ -529,7 +594,7 @@ fu_synaptics_rmi_firmware_parse(FuFirmware *firmware,
 				    error))
 		return FALSE;
 	fu_firmware_set_size(firmware, firmware_size);
-
+	g_debug("bootloader version is %d \n", self->bootloader_version);
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c:597]: why this? isn't it in the to_string() output as well?

File: plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c [hunk 3]
```diff
@@ -574,6 +640,40 @@ fu_synaptics_rmi_firmware_get_sig_size(FuSynapticsRmiFirmware *self)
 	return self->sig_size;
 }
 
+GPtrArray *
+fu_synaptics_rmi_firmware_get_image_by_id(FuFirmware *firmware, const gchar *id)
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c:644]: fu_synaptics_rmi_firmware_get_image**s**_by_id no?

File: plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c [hunk 4]
```diff
@@ -574,6 +640,40 @@ fu_synaptics_rmi_firmware_get_sig_size(FuSynapticsRmiFirmware *self)
...
+{
+	g_autoptr(GPtrArray) imgs = NULL;
+	g_autoptr(GBytes) img = NULL;
+	g_autoptr(GBytes) img_sig = NULL;
+	g_autoptr(GError) error_local = NULL;
+	imgs = g_ptr_array_new();
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c:650]: just do this inline above

File: plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c [hunk 5]
```diff
@@ -574,6 +640,40 @@ fu_synaptics_rmi_firmware_get_sig_size(FuSynapticsRmiFirmware *self)
...
+
+	img = fu_firmware_get_image_by_id_bytes(firmware, id, &error_local);
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c:652]: you need to handle the error

File: plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c [hunk 6]
```diff
@@ -574,6 +640,40 @@ fu_synaptics_rmi_firmware_get_sig_size(FuSynapticsRmiFirmware *self)
...
+	g_ptr_array_add(imgs, img);
+	img_sig = fu_firmware_get_image_by_id_bytes(firmware,
+						    g_strdup_printf("%s-signature", id),
+						    &error_local);
+	g_ptr_array_add(imgs, img_sig);
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c:657]: and handle the error here too

File: plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c [hunk 7]
```diff
@@ -326,8 +321,58 @@ fu_synaptics_rmi_v7_device_write_partition(FuSynapticsRmiDevice *self,
 							     fu_chunk_get_data_sz(chk),
 							     error))
 			return FALSE;
-		fu_progress_step_done(progress);
 	}
+
+	/* write signature */
+	if (g_ptr_array_index(bytes, 1)) {
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c:327]: this is odd; do you mean `bytes->len > 1`? It's also weird to call it bytes if it's an array.

File: plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c [hunk 8]
```diff
@@ -326,8 +321,58 @@ fu_synaptics_rmi_v7_device_write_partition(FuSynapticsRmiDevice *self,
...
+		fu_byte_array_append_uint16(req_offset, 0x0, G_LITTLE_ENDIAN);
+		if (!fu_synaptics_rmi_device_write(self,
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c:329]: i think all of this signature chunk should be split out as a new function

File: plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c [hunk 9]
```diff
@@ -326,8 +321,58 @@ fu_synaptics_rmi_v7_device_write_partition(FuSynapticsRmiDevice *self,
...
+							   error)) {
+				g_prefix_error(error, "failed to flash command: ");
+				return FALSE;
+			}
+			if (!fu_synaptics_rmi_v7_device_write_blocks(self,
+								     f34->data_base + 0x5,
+								     fu_chunk_get_data(chk),
+								     fu_chunk_get_data_sz(chk),
+								     error))
+				return FALSE;
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c:372]: I think you want to update the progress for each chunk here.

File: plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c [hunk 10]
```diff
@@ -435,6 +521,86 @@ typedef struct __attribute__((packed)) {
...
+	FuSynapticsRmiFlash *flash = fu_synaptics_rmi_device_get_flash(self);
+	FuSynapticsRmiFunction *f34;
+	g_autoptr(GByteArray) req_addr_zero = g_byte_array_new();
+	g_autoptr(GByteArray) req_cmd = g_byte_array_new();
+	g_autoptr(GByteArray) req_partition_id = g_byte_array_new();
+	g_autoptr(GByteArray) req_transfer_length = g_byte_array_new();
+	g_autoptr(GByteArray) res = NULL;
+	g_autoptr(GByteArray) pubkey = g_byte_array_new();
+	gsize key_size = 0x100;
+	/* f34 */
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c:536]: need a newline before this

File: plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c [hunk 11]
```diff
@@ -435,6 +521,86 @@ typedef struct __attribute__((packed)) {
...
+	/* read back entire buffer in blocks */
+	res = fu_synaptics_rmi_device_read(self, f34->data_base + 0x5, (guint32)key_size, error);
+	if (res == NULL) {
+		g_prefix_error(error, "failed to read: ");
+		return NULL;
+	}
+
+	for (guint i = 0; i < res->len; i++) {
+		fu_byte_array_append_uint8(pubkey, res->data[res->len - i - 1]);
+	}
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c:598]: no need for braces

File: plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c [hunk 12]
```diff
@@ -514,12 +681,15 @@ fu_synaptics_rmi_device_read_flash_config_v7(FuSynapticsRmiDevice *self, GError
 			     "FlashConfig",
 			     res->data,
 			     res->len,
-			     80,
+			     40,
 			     FU_DUMP_FLAGS_NONE);
 	}
+	if ((res->data[0] & 0x0f) == 1) {
+		partition_size = sizeof(RmiPartitionTbl) + 2;
+	}
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c:689]: no need for braces

File: plugins/synaptics-rmi/fu-synaptics-rmi-device.c [hunk 13]
```diff
@@ -622,14 +621,20 @@ fu_synaptics_rmi_device_poll(FuSynapticsRmiDevice *self, GError **error)
...
-		g_set_error(error,
-			    FWUPD_ERROR,
-			    FWUPD_ERROR_WRITE,
-			    "flash status invalid: 0x%x",
-			    (guint)(f34_db->data[0] & 0x1f));
+		if ((f34_db->data[0] & 0x1f) == 0x6) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_FOUND,
+				    "unknown erase or program, expected during erase");
```
[the remark this answers is not in the document]
Comment (vhuag) [reply plugins/synaptics-rmi/fu-synaptics-rmi-device.c:628]: This is for erase function to ignore this error when the partition is not found, the erase_ap command may not erase some partition so we want to make sure it's clear.

File: plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c (hunk 1 above)
Comment (vhuag) [reply plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c]: Fixed.

File: plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c (hunk 2 above)
Comment (vhuag) [reply plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c]: Removed.

File: plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c (hunk 4 above)
Comment (vhuag) [reply plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c]: Fixed.

File: plugins/synaptics-rmi/fu-synaptics-rmi-device.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (hughsie) [reply plugins/synaptics-rmi/fu-synaptics-rmi-device.c:628]: right, i just think this is easier to read:

	if ((f34_db->data[0] & 0x1f) == 0x6) {
		g_set_error(error,
			    FWUPD_ERROR,
			    FWUPD_ERROR_NOT_FOUND,
			    "unknown erase or program, expected during erase");
	} else {
		g_set_error(error,
			    FWUPD_ERROR,
			    FWUPD_ERROR_WRITE,
			    "flash status invalid: 0x%x",
			    (guint)(f34_db->data[0] & 0x1f));
	}
	return FALSE;

to...

	if ((f34_db->data[0] & 0x1f) == 0x6) {
		g_set_error(error,
			    FWUPD_ERROR,
			    FWUPD_ERROR_NOT_FOUND,
			    "unknown erase or program, expected during erase");
		return FALSE;
	}
	g_set_error(error,
		    FWUPD_ERROR,
		    FWUPD_ERROR_WRITE,
		    "flash status invalid: 0x%x",
		    (guint)(f34_db->data[0] & 0x1f));
	}
	return FALSE;

File: plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c (hunk 5 above)
Comment (hughsie) [reply plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c:651]: I'd be much happier if fu_synaptics_rmi_firmware_get_image_by_id() had a GError parameter...

File: plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c [hunk 14]
```diff
@@ -339,20 +384,25 @@ fu_synaptics_rmi_v7_device_write_firmware(FuDevice *device,
...
 	FuSynapticsRmiFunction *f34;
-	g_autoptr(GBytes) bytes_bin = NULL;
-	g_autoptr(GBytes) bytes_cfg = NULL;
-	g_autoptr(GBytes) bytes_flashcfg = NULL;
+	g_autoptr(GPtrArray) bytes_bin = NULL;
+	g_autoptr(GPtrArray) bytes_cfg = NULL;
+	g_autoptr(GPtrArray) bytes_flashcfg = NULL;
+	g_autoptr(GPtrArray) bytes_fld = NULL;
+	g_autoptr(GPtrArray) bytes_afe = NULL;
+	g_autoptr(GPtrArray) bytes_displaycfg = NULL;
```
[the remark this answers is not in the document]
Comment (vhuag) [reply plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c]: fixed.

File: plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c (hunk 10 above)
Comment (vhuag) [reply plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c:537]: fixed.

File: plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c (hunk 12 above)
Comment (vhuag) [reply plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c]: removed.

File: plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c [hunk 15]
```diff
@@ -722,3 +895,42 @@ fu_synaptics_rmi_v7_device_query_status(FuSynapticsRmiDevice *self, GError **err
...
+fu_synaptics_rmi_v7_device_secure_check(FuSynapticsRmiDevice *self,
+					FuFirmware *firmware,
+					GError **error)
+{
+	g_autoptr(GBytes) pubkey = NULL;
+	gboolean verify_result = FALSE;
+	g_autoptr(GPtrArray) imgs = NULL;
+	FuSynapticsRmiFlash *flash = fu_synaptics_rmi_device_get_flash(self);
+
+	if (flash->bootloader_id[0] < 10 && flash->has_pubkey) {
```
[the remark this answers is not in the document]
Comment (vhuag) [reply plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c]: changed.

File: plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c [hunk 16]
```diff
@@ -169,6 +170,33 @@ fu_synaptics_rmi_firmware_add_image(FuFirmware *firmware,
...
+					const gchar *id,
+					GBytes *fw,
+					gsize offset,
+					gsize sz,
+					gsize sig_sz,
+					GError **error)
+{
+	g_autoptr(GBytes) bytes = NULL;
+	g_autoptr(FuFirmware) img = NULL;
+	g_autofree gchar *sig_id = NULL;
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-firmware.c:184]: I think you can move these into the `if (sig_sz` scope.

File: plugins/synaptics-rmi/fu-synaptics-rmi-hid-device.c [hunk 17]
```diff
@@ -392,8 +393,23 @@ fu_synaptics_rmi_hid_device_rebind_driver(FuSynapticsRmiDevice *self, GError **e
...
 
+	/* build paths */
+	parent_phys = g_udev_device_get_parent_with_subsystem(udev_device, "i2c", NULL);
+	if (parent_phys == NULL) {
+		parent_phys = g_udev_device_get_parent_with_subsystem(udev_device, "usb", NULL);
+		if (parent_phys == NULL) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "no I2C parent device for %s",
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-hid-device.c:404]: `no I2C or USB parent device` perhaps?

File: plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c [hunk 18]
```diff
@@ -117,32 +123,26 @@ fu_synaptics_rmi_v7_device_detach(FuDevice *device, FuProgress *progress, GError
...
-fu_synaptics_rmi_v7_device_erase_all(FuSynapticsRmiDevice *self, GError **error)
+fu_synaptics_rmi_v7_device_erase_partition(FuSynapticsRmiDevice *self,
+					   guint8 partition_id,
+					   guint cmd,
+					   GError **error)
 {
 	FuSynapticsRmiFunction *f34;
 	FuSynapticsRmiFlash *flash = fu_synaptics_rmi_device_get_flash(self);
 	g_autoptr(GByteArray) erase_cmd = g_byte_array_new();
-
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c:125]: keep this newline pls

File: plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c [hunk 19]
```diff
@@ -152,54 +152,47 @@ fu_synaptics_rmi_v7_device_erase_all(FuSynapticsRmiDevice *self, GError **error)
...
-						   FU_SYNAPTICS_RMI_DEVICE_FLAG_NONE,
-						   error)) {
-			g_prefix_error(error, "failed to erase core config: ");
-			return FALSE;
-		}
+	if (!fu_synaptics_rmi_v7_device_erase_partition(self,
+							RMI_PARTITION_ID_CORE_CODE,
+							RMI_FLASH_CMD_ERASE_AP,
+							&error_local)) {
+		return FALSE;
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c:183]: you mean `error`, right?

File: plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c [hunk 20]
```diff
@@ -326,8 +321,58 @@ fu_synaptics_rmi_v7_device_write_partition(FuSynapticsRmiDevice *self,
...
+		if (!fu_synaptics_rmi_device_write(self,
+						   f34->data_base + 0x2,
+						   req_offset,
+						   FU_SYNAPTICS_RMI_DEVICE_FLAG_NONE,
+						   error)) {
+			g_prefix_error(error, "failed to write offset: ");
+			return FALSE;
+		}
+
+		chunks = fu_chunk_array_new_from_bytes(g_ptr_array_index(array_img, 1),
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c:338]: yup, you're leaking memory -- splitting this out will allow you to define your "own" chunks.

File: plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c [hunk 21]
```diff
@@ -368,58 +418,94 @@ fu_synaptics_rmi_v7_device_write_firmware(FuDevice *device,
...
-			return FALSE;
-	}
 
 	/* disable powersaving */
 	if (!fu_synaptics_rmi_device_disable_sleep(self, error))
 		return FALSE;
 	fu_progress_step_done(progress);
 
+	/* write fld before erase */
+	if (g_ptr_array_index(array_fld, 0) != NULL) {
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c:437]: array_fld->len pls

File: plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c [hunk 22]
```diff
@@ -435,6 +521,86 @@ typedef struct __attribute__((packed)) {
...
+{
+	FuSynapticsRmiFlash *flash = fu_synaptics_rmi_device_get_flash(self);
+	FuSynapticsRmiFunction *f34;
+	g_autoptr(GByteArray) req_addr_zero = g_byte_array_new();
+	g_autoptr(GByteArray) req_cmd = g_byte_array_new();
+	g_autoptr(GByteArray) req_partition_id = g_byte_array_new();
+	g_autoptr(GByteArray) req_transfer_length = g_byte_array_new();
+	g_autoptr(GByteArray) res = NULL;
+	g_autoptr(GByteArray) pubkey = g_byte_array_new();
+	gsize key_size = 0x100;
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c:535]: const?

File: plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c [hunk 23]
```diff
@@ -722,3 +894,43 @@ fu_synaptics_rmi_v7_device_query_status(FuSynapticsRmiDevice *self, GError **err
...
 }
+
+gboolean
+fu_synaptics_rmi_v7_device_secure_check(FuSynapticsRmiDevice *self,
+					FuFirmware *firmware,
+					GError **error)
+{
+	g_autoptr(GBytes) pubkey = NULL;
+	gboolean verify_result = FALSE;
+	g_autoptr(GPtrArray) imgs = NULL;
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c:905]: normally the g_autoptr's are sorted and last in the block.

File: plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c [hunk 24]
```diff
@@ -722,3 +894,43 @@ fu_synaptics_rmi_v7_device_query_status(FuSynapticsRmiDevice *self, GError **err
...
+	FuSynapticsRmiFlash *flash = fu_synaptics_rmi_device_get_flash(self);
+
+	if (flash->bootloader_id[0] >= 10 || flash->has_pubkey == FALSE)
+		return TRUE;
+
+	imgs = fu_synaptics_rmi_firmware_get_images(firmware);
+	pubkey = fu_synaptics_rmi_v7_device_get_pubkey(self, error);
+	if (pubkey == NULL) {
+		g_prefix_error(error, "get pubkey failed.");
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c:914]: g_prefix_error needs to have a trailing `: `

File: plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c [hunk 25]
```diff
@@ -722,3 +894,43 @@ fu_synaptics_rmi_v7_device_query_status(FuSynapticsRmiDevice *self, GError **err
...
+			verify_result = fu_synaptics_verify_sha256_signature(
+			    g_ptr_array_index(array_payload, 0),
+			    pubkey,
+			    g_ptr_array_index(array_payload, 1),
+			    error);
+			if (verify_result < 0) {
+				g_prefix_error(error, "%s secure check failed. ", id);
+				return FALSE;
+			}
+			g_debug("%s signature verified successfully \n", id);
```
Comment (hughsie) [inline plugins/synaptics-rmi/fu-synaptics-rmi-v7-device.c:932]: no `\n` in g_debug's please

**Comment by vhuag** [comment]
Hi Richard.
I'll close this pull request and submit the each feature separately to make things more clear.
Thanks

**Pull request closed without merging**